balance_service/enforcement/usage_enforcement.py

```diff
@@ -0,0 +1,322 @@
+                ),
+                end_time=self._convert_to_localtime(
+                    self._date_from_string(lease["end_date"])
+                ),
+                hourly_cost=self.__get_reservation_sus(
+                    reservation["resource_type"], reservation["allocations"]
+                ),
+            )
+            new_charge.save()
+
+    def check_usage_against_allocation_update(self, data):
+        """Check if we have enough available SUs for update"""
+        context = data["context"]
+        old_lease = data["current_lease"]
+        new_lease = data["lease"]
+
+        old_lease_eval = self.evaluate_lease(context, old_lease)
+        new_lease_eval = self.evaluate_lease(context, new_lease)
+
+        LOG.debug(
+            (
+                "Evaluating lease update request: "
+                f"old_lease={old_lease_eval}, new_lease={new_lease_eval}"
+            )
+        )
+        debug_lease_reservations("Old reservation", old_lease)
+        debug_lease_reservations("New reservation", new_lease)
+
+        left = self.get_remaining_balance(old_lease_eval.project.id)
+
+        change_amount = new_lease_eval.amount - old_lease_eval.amount
+
+        if change_amount > left:
+            raise exceptions.BillingError(
+                "Reservation update would spend {:.2f} more SUs, only "
+                "{:.2f} left".format(change_amount, left)
+            )
+
+        # create/update charges
+        now = timezone.now()
+        end_date_changed = old_lease["end_date"] != new_lease["end_date"]
+        for reservation in new_lease["reservations"]:
+            new_hourly_cost = self.__get_reservation_sus(
+                reservation["resource_type"], reservation["allocations"]
+            )
+            if not end_date_changed:
+                # check if hourly_cost changed
+                old_reservation = [
+                    r for r in old_lease["reservations"] if r["id"] == reservation["id"]
+                ]
+                old_hourly_cost = None
+                if len(old_reservation) > 0:
+                    old_reservation = old_reservation[0]
+                    old_hourly_cost = self.__get_reservation_sus(
+                        old_reservation["resource_type"], old_reservation["allocations"]
+                    )
+                if new_hourly_cost == old_hourly_cost:
+                    # nothing changed
+                    continue
+            reservation_charges = self._get_charges_by_reservation(
+                reservation["id"], new_lease_eval.region
+            )
+
+            # get the ongoing charnge
+            ongoing_charge = [c for c in reservation_charges if c.end_time > now]
+            # should have exactly one ongoing charge
+            if len(ongoing_charge) == 1:
+                ongoing_charge = ongoing_charge[0]
+                ongoing_charge.end_time = now
+                ongoing_charge.save()
+            else:
+                LOG.error(
```

Comment:
Is it appropriate to raise an exception here, do you think? Should this just proceed w/o an error if there are multiple ongoing charges for the reservation somehow? Currently the function will proceed, which will just make the problem worse by continuing to add more ongoing charges.